### ISSUE TITLE
Model robustness

### ISSUE THREAD

**Issue by lghasemzadeh** [vladmandic/human#63]
Hello,
I have checked the facemesh demo of Mediapipe (https://google.github.io/mediapipe/solutions/face_mesh) and face landmark detection (https://blog.tensorflow.org/2020/11/iris-landmark-tracking-in-browser-with-MediaPipe-and-TensorFlowJS.html). The second one really works better and more robust even when I change my head pose extremely and quickly.
1. But which one did you use in your model? If you used the second one why it is not robust and the facemesh disappears suddenly when I change my head pose extremely? 
2. hand tracking is not working for both hands. Is there a reason for that since in the mediapipe's demo it works very well?
3. For the age prediction which model did you used? a pre-trained network? 
4. For mood prediction, did you used deep leaning model or just it takes the differences between nodes around eyebrow, lips and eyes?

Thank you

**Comment by vladmandic** [comment]
> But which one did you use in your model? If you used the second one why it is not robust and the facemesh disappears suddenly when I change my head pose extremely?

it's a tradeoff between precision and performance and it's fully tunable in configuration.
how it works is that library tries to re-use previously detected face bounding box without running it again and just runs the mesh and iris detector instead. so if the face moves suddenly, face will disappear and bounding box gets invalidated so it re-runs on the next frame. again, behavior is fully tunable.

also disabled by default for performance reasons is face angle calculation so if head tilts more than ~20 degrees, it will get invalidated. also can be enabled via configuration.

> hand tracking is not working for both hands. Is there a reason for that since in the mediapipe's demo it works very well?

yes, it does, you just have to set configuration parameter `hand.maxHands` to a value higher than 1. again, it's about performance vs precision.

> For the age prediction which model did you used? a pre-trained network?

Pre-trained network, trained on imdb dataset.

> For mood prediction, did you used deep leaning model or just it takes the differences between nodes around eyebrow, lips and eyes?

pre-trained network, not just simple math.  
math is used to calculate gestures such as `mouth open, raise left hand, looking down, looking at camera, etc.`.

list of all models used are in credits page: <https://github.com/vladmandic/human/wiki/Models>  
list of pre-defined math-based gestures that are included (others can be added by user) is in <https://github.com/vladmandic/human/wiki/Gesture>

**Comment by lghasemzadeh** [comment]
Thank you very much for the detailed answer.
I still have lots of questions about mediapipe also. I don't want to bother you, so is there any forum for asking these questions?

**Comment by vladmandic** [comment]
for mediapipe, author state that general questions should be posted via <https://stackoverflow.com/questions/tagged/mediapipe>

**Comment by vladmandic** [comment]
closing the issue as the original question is answered.

**Comment by lghasemzadeh** [comment]
Hello Vladimir,

I am sorry for asking alot, I am amateur but very curious about this library.

What do you mean by performance? precision is also a concept of performance. you mean if I make the model to act more accurate, the performance will decrease? but what is performance here? is it the speed of algorithm?

mediapipe (the new version) also does the face detection/bounding box/mesh/iris jobs as well but why it doesn't have lag, shaking and disappearing even when I rotate my head more that 30 or 40 degrees? in its demo I don't tune anything and it works very well.

 For the age prediction have you done any fine tuning over the pre-trained network or any other changes?

For emotion detection, why you didn't use simple math (calculating difference in distances between eyebrow etc.) what will be its deficiency or problem?

If simple math won't work for emotion detection correctly, would it work for mouth opening and gaze direction?!!

Thx
@vladmandic

**Comment by vladmandic** [comment]
>you mean if I make the model to act more accurate, the performance will decrease? but what is performance here? is it the speed of algorithm?

yes, if you make model to act more accurately it will decrease performanse. this can be due to many things.
for example (and there are other examples as well):
- input resolution: higher resolution, better accuracy but more data has to be processed by model so it's slower.
- pre&post processing: for example, after face detect model returns bounding box, there is additional math to calculate the rotation angle of the face and rotate the image within bounding box before passing it to face mesh model. but that rotating image is computationally intensive, so it's optional - it increases precision, but decreases performance.
- how often are some steps skipped: i've mentioned that face detect does not run on every frame by default, my library tries to re-use previously detected bounding box to skip re-running that model before passing bounding box to face mesh. but if face moves too fast, it will get out of frame. optimally, you can run face detect on each frame, but it's slower.

> mediapipe (the new version) also does the face detection/bounding box/mesh/iris jobs as well but why it doesn't have lag, shaking and disappearing even when I rotate my head more that 30 or 40 degrees?  

new version of mediapipe is written in C++ and compiled to WASM. so entire pre and post processing that i've mentioned above is much faster than in JavaScript.  

also, JavaScript is single threaded, so any calculation will block processing - thus more lag. mediapipe can perform asynchronous interpolations between model runs. meaning, it runs the model prediction, but also calculates averages between each run so it can draw much smoother frames.  

i have added async processing in `human` library as well, but due to overal nature of JavaScript, it doesn't help much.

in reality, mediapipe is a superior way of doing it, but it's proprietary - you can only do what they allow you and not much else, not room for user modifications/improvements/tunings.  

also, mediapipe models are pre-compiled to tflite format and cannot be easily re-used elsewhere.

> For the age prediction have you done any fine tuning over the pre-trained network or any other changes?

no additional fine tuning.

> For emotion detection, why you didn't use simple math (calculating difference in distances between eyebrow etc.) what will be its deficiency or problem?

it's possible, but emotions are complex - it's easy to differentiate between angry vs happy, but not so easy to mathematically describe surprise vs digust. i just found using model to be easier and more reliable.

> If simple math won't work for emotion detection correctly, would it work for mouth opening and gaze direction?!!

it does work for anything you can clearly describe mathematically. so things like 'mouth open' or 'looking left' or 'leaning down' are easy.

**Comment by lghasemzadeh** [comment]
The other reason of lower performance can be merging several models together in human, is it correct? since mediapipe has separated all functions (e.g. hand tracking and facemesh are separated and work solely, maybe that is why it has better accuracy and performance?)

Do have plan to publish the python version of your work?

The demo that I checked was for facemesh and iris tracking which is provided for js! not C++ I think
https://blog.tensorflow.org/2020/11/iris-landmark-tracking-in-browser-with-MediaPipe-and-TensorFlowJS.html
![Screenshot from 2021-01-28 22-31-37](https://user-images.githubusercontent.com/36885164/106188926-a54b8a00-61b8-11eb-8a68-f436c57bfd4f.png)
If I'm not mistaken the older version has Iris and mesh for C++,

Since I just focus on eyes and head, is it better to use mediapipe not Human? since I need both high accuracy and performance at the same time.

I previously wrote a script which was using dlib and tensorflow facial landmarks, from those landmarks I calculated for head pose and gaze direction (a rule based/mathematical method, not using any DL) but since the landmarks were not accurate, my model had low accuracy. since this facemesh is very precise about coordinate of eyes/Iris landmarks, I think it would solve my problem.
I saw this solution from this video in linkedin which is similar to your package: https://www.linkedin.com/feed/update/urn:li:activity:6730449764283895810/

In above you mentioned I can easily mathematical calculate for looking left or down, are they just examples and is it true for other directions such as up and right?

**Comment by vladmandic** [comment]
> The other reason of lower performance can be merging several models together in human, is it correct?

All models in `Human` are separate and each can be enabled/disabled individually in configuration.
There are some dependencies such as:
- FaceMesh, Iris, Gender, Age, Emotion depend on FaceDetect as they all work on cropped image of a face
- HandPose depends on HandDetect
- BodyPose depends on BodyDetect

> Do have plan to publish the python version of your work?

No.

> The demo that I checked was for facemesh and iris tracking which is provided for js! not C++ I think

MediaPipe JS solution uses JS just to load WASM module which was written in C++. Actual processing is done in WASM, not JS.

> Since I just focus on eyes and head, is it better to use mediapipe not Human? since I need both high accuracy and performance at the same time.

Your choice. With `Human` you get full configurability - you can tweak it as much as you want to get equal or even higher precision than MediaPipe. On the other hand, using MediaPipe out-of-the-box is pretty good.

> In above you mentioned I can easily mathematical calculate for looking left or down, are they just examples and is it true for other directions such as up and right?

Any such rules are simple (and yes, up and down are already included). Complex rules that cannot be done reliably with just math and where pre-trained models are better would be ones that vary from face to face - for example, your facial expression for surprise might not be mathematically same as for me.

**Comment by lghasemzadeh** [comment]
ok, thank you

Is this library (Human) available for both academic and industry use? I know it is open-source but is it for both purpose freely?

**Comment by vladmandic** [comment]
Re: license

Library is released under MIT license.
Some models are originally under Apache 2.0 license.

Both are permissive - so in short, yes.

**Comment by lghasemzadeh** [comment]
I really appreciate your detailed answer.

In the Human library is there any black box module? can I use/change/modify all the functions and modules?

**Comment by ButzYung** [comment]
In terms of functionality, `human`/TFJS solution is obviously better than the current MediaPipe JS soltuion. The current MediaPipe JS solution is quite buggy. It doesn't work on web worker. It is not well documented and sometimes you have to do trial-n-error just to make things work. Its output is bugged. For example, MediaPipe JS facemesh output doesn't give the z coordinates for some unknown reasons.

**Comment by vladmandic** [comment]
@lghasemzadeh other than the fact that included models are pre-trained, there is nothing you cannot change/modify.
and if you want to add additional modules or swap out existing ones, that is also pretty easy as design is modular.

**Comment by lghasemzadeh** [comment]
Hello Vladimir,

Hope you are well.
1) It seems that the Iris direction (Gaze Direction) is not working, it always print 'Looking at Camera' while I am completely looking at Left or Right. Is there any way to correct or improve it?
2) The head pose works very well but there are two face description on the screen, one for left and right, and another for up and down. what is the reason to separate them? 

Thank you

**Comment by vladmandic** [comment]
@lghasemzadeh 

> Hope you are well.

Yes, thank you! :-)

Btw, can you open a new issue for this? Both questions are valid, but unrelated to this issue that is already closed.

> It seems that the Iris direction (Gaze Direction) is not working, it always print 'Looking at Camera' while I am completely looking at Left or Right. Is there any way to correct or improve it?

I just tried using webcam and it works for me.  
Can you send some sample images where its misdetected along with details of your configuration?

Note that mathematically 'looking at camera' is defined as difference between area size of left and right iris and if they are within 25% (see `src/gestures/gestures.ts:iris()`, then `human` will return `looking at camera`. This threshold can be changed from 25% to anything, but it's not exposed as user-configurable value as I don't want to have too complex of configuration file (it's already pretty complex), so I just use some value that looked ok.

> The head pose works very well but there are two face description on the screen, one for left and right, and another for up and down. what is the reason to separate them?

Face can be up/center/down and left/center/right at the same time, so gestures returns two separate results.
But gestures return object is just that - an object - and If you want to combine them, it's just a question of printing them together instead separately.

**Comment by lghasemzadeh** [comment]
Ok, I will open a new issue
